### ISSUE TITLE
Mox.stub_with/2 does not work with skip_optional_callbacks

### ISSUE THREAD

**Issue by flurin** [dashbitco/mox#95]
Using `Mox.stub_with/2` in combination with skipping optional callbacks will result into an `(ArgumentError) unknown function XXX for mock XXX` error. Looking at the Mox code for `stub_with` it will take all callbacks instead of skipping those as defined in `skip_optional_callbacks`. 

A simple way to reproduce this:

```
defmodule MyBehaviour do
  @callback always(any()) :: any()
  @callback optional(any()) :: any()

  @optional_callbacks optional: 1
end

defmodule MyStub do
  @behaviour MyBehaviour
  def always(_), do: true
end

Mox.defmock(
  MyMock,
  for: MyBehaviour,
  skip_optional_callbacks: true
)
```

Doing:

```
MyMock |> stub_with(MyStub)
```

In a test will result in the following error:

```
 ** (ArgumentError) unknown function optional/1 for mock MyMock
     code: MyMock |> stub_with(MyStub)
     stacktrace:
       (mox 0.5.2) lib/mox.ex:546: Mox.add_expectation!/4
       (mox 0.5.2) lib/mox.ex:477: Mox.stub/3
       (mox 0.5.2) lib/mox.ex:528: anonymous fn/4 in Mox.stub_with/2
       (elixir 1.10.3) lib/enum.ex:2111: Enum."-reduce/3-lists^foldl/2-0-"/3
       (mox 0.5.2) lib/mox.ex:525: Mox.stub_with/2
       test/helper/my_test.exs:8: (test)
```

**Comment by josevalim** [comment]
Good catch. A PR is welcome!

**Comment by flurin** [comment]
@josevalim I would love to, but I'm not entirely sure what direction to take. Currently it just iterates over everything from the behaviour: `behaviour.behaviour_info(:callbacks)`, my gut feeling says it would be better to only iterate over the functions that actually have mock functions defined by using `behaviour.__info__(:functions)` but then we'd need to filter out `__mock_for__`. 

What do you think?

**Comment by josevalim** [comment]
@flurin We will need to see if the function is defined in the stub and if it isn't and the callback is optional, we skip it instead of raising.

**Comment by flurin** [comment]
Ok! I'll make a PR in the coming days.

**Comment by flurin** [comment]
The fix in #96 checks wether or not there is a mock function defined and only stubs it when it is. Moving the check to where we raise would require passing the stub module around and I thought that would be less then ideal.

**Comment by josevalim** [comment]
Closing in favor of the PR.
